### ISSUE TITLE
Fix TestSourceCatalogObjectHGPS.test_ecpl_model fail

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#687]
This is a reminder issue for myself to fix this test fail I'm seeing locally.
(this is a test that doesn't run on travis-ci because the data isn't public yet)

```
________________________________________________________________ TestSourceCatalogObjectHGPS.test_ecpl_model ________________________________________________________________

self = <gammapy.catalog.tests.test_hess.TestSourceCatalogObjectHGPS object at 0x12363f748>

    def test_ecpl_model(self):
        model = self.cat['HESS J0835-455'].spectral_model
        pars = model.parameters
        assert_quantity_allclose(pars.amplitude, Quantity(6.408420542586617e-12, 'TeV-1 cm-2 s-1'))
        assert_quantity_allclose(pars.index, Quantity(1.3543991614920847, ''))
        assert_quantity_allclose(pars.reference, Quantity(1.696938754239, 'TeV'))
        assert_quantity_allclose(pars.lambda_, Quantity(0.081517637, 'TeV-1'))

        emin, emax = Quantity([1, 1E10], 'TeV')
        desired =  Quantity(self.source.data['Flux_Spec_PL_Int_1TeV'], 'cm-2 s-1')
>       assert_quantity_allclose(model.integral(emin, emax), desired, rtol=0.01)

gammapy/catalog/tests/test_hess.py:109: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
gammapy/spectrum/models.py:65: in integral
    return integrate_spectrum(self, emin, emax, **kwargs)
gammapy/spectrum/utils.py:285: in integrate_spectrum
    val = _trapz_loglog(y, x, ulog10=True, intervals=intervals)
gammapy/spectrum/utils.py:350: in _trapz_loglog
    b = log10(y[slice2] / y[slice1]) / log10(x[slice2] / x[slice1])
/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/numpy/lib/function_base.py:2218: in __call__
    return self._vectorize_call(func=func, args=vargs)
/opt/local/Library/Frameworks/Python.framework/Versions/3.5/lib/python3.5/site-packages/numpy/lib/function_base.py:2287: in _vectorize_call
    outputs = ufunc(*inputs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (0.0,), kwargs = {}, pos_w_uncert = [], names_w_uncert = set()

    @set_doc("""\
        Version of %s(...) that returns an affine approximation
        (AffineScalarFunc object), if its result depends on variables
        (Variable objects).  Otherwise, returns a simple constant (when
        applied to constant arguments).

        Warning: arguments of the function that are not AffineScalarFunc
        objects must not depend on uncertainties.Variable objects in any
        way.  Otherwise, the dependence of the result in
        uncertainties.Variable objects will be incorrect.

        Original documentation:
        %s""" % (f.__name__, f.__doc__))
    def f_with_affine_output(*args, **kwargs):

        ########################################
        # The involved random variables must first be gathered, so
        # that they can be independently updated.

        # The arguments that contain an uncertainty (AffineScalarFunc
        # objects) are gathered, as positions or names; they will be
        # replaced by simple floats.

        pos_w_uncert = [index for (index, value) in enumerate(args)
                        if isinstance(value, AffineScalarFunc)]
        names_w_uncert = set(key for (key, value) in kwargs.items()
                             if isinstance(value, AffineScalarFunc))

        ########################################
        # Value of f() at the nominal value of the arguments with
        # uncertainty:

        # The usual behavior of f() is kept, if no number with
        # uncertainty is provided:
        if (not pos_w_uncert) and (not names_w_uncert):
>           return f(*args, **kwargs)
E           ValueError: math domain error

/Users/deil/Library/Python/3.5/lib/python/site-packages/uncertainties/core.py:621: ValueError
```

**Comment by adonath** [comment]
I've looked into that issue and I can reproduce it on my machine. The reason of this error is the upper bound of the integration range (emax = `1E10 * u.TeV`). Effectively the function value  `model(1E3 * u.TeV)` is already O(1E-51) (because of the exponential term...), so `1E10` is far beyond the machine precision, maybe one should clip the values somehow or just change the integration range to `emax = 1E3 * u.TeV`.

**Comment by cdeil** [comment]
@adonath - I don't fully understand. There will probably be Fermi-LAT pulsars with a sharp cutoff at ~ 1 GeV that have `flux < 1e-51` at `1e3 TeV`, so changing the `e_max` to fix this test fail will probably just mean that the same issue will come back for a different case soon? We need an integration method that's OK with fluxes ~ 0, no?

That said -- if you have any solution to fix / improve this now, please just commit it, and maybe leave a `TODO` marker or reminder issue for the 0.6 milestone.

**Comment by adonath** [comment]
@cdeil Yes, you are right this wasn't the full issue. I tracked the issue down to the `uncertainties.unumpy.log10` function. While numpy deals fine with these small numbers:
```
>>> import numpy as np
>>> np.log10(1E-500)
-inf
```
The invalid numbers are then handled in the [this line](https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/utils.py#L291) in `_trapz_loglog`.

The `uncertainties.unumpy.log10` raises an error:
```
>>> from uncertainties.unumpy import log10
>>> log10(1E-500)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/hfm/adonath/.local/lib/python2.7/site-packages/numpy/lib/function_base.py", line 2218, in __call__
    return self._vectorize_call(func=func, args=vargs)
  File "/home/hfm/adonath/.local/lib/python2.7/site-packages/numpy/lib/function_base.py", line 2287, in _vectorize_call
    outputs = ufunc(*inputs)
  File "/home/hfm/adonath/.local/lib/python2.7/site-packages/uncertainties/__init__.py", line 845, in f_with_affine_output
    return f(*args, **kwargs)
ValueError: math domain error
```

And the reason is, that `uncertainties.unumpy.log10` doesn't call into `np.log10`, but rather a vectorized version of `math.log10` (see [here](https://github.com/lebigot/uncertainties/blob/master/uncertainties/unumpy/core.py#L644)), which wasn't clear to me.

The default behaviour is to use `uncertainties.unumpy.log10`, which should be changed. But still we need some kind of fix for the use where we want to integrate with uncertainties. Do you have a suggestion?

**Comment by adonath** [comment]
One could manually filter out the tiny values like so:
```
import numpy as np
y = np.array([1E-300, 1E-500, 1E-800])
valid = y > np.finfo(float).tiny
```
@cdeil Do think this is appropriate?

**Comment by cdeil** [comment]
Sure, if that works, and you leave a comment why you're doing it, it's certainly fine with me.
Maybe call [numpy.clip](https://docs.scipy.org/doc/numpy/reference/generated/numpy.clip.html) to clip the lower fluxes to a range that works?
(not sure if that's better than filtering, as you like)

**Comment by cdeil** [comment]
@adonath - I think this issue was fixed in #775, no? Can we close it?

**Comment by adonath** [comment]
Yes
